### ISSUE TITLE
New event that gets triggered when a player moves

### ISSUE THREAD

**Issue by evi1a** [haxball/haxball-issues#468]
Since the player's position doesn't get automatically updated without using onGameTick and checking all players all the time, I would like to request a very useful event, that gets triggered when a player moves (no matter if he moves by himself or if another entity causes him to move).

`onPlayerUpdatePosition(player : PlayerObject) : void`
Event called when a player moves.

And maybe to make it more complete:

`onBallUpdatePosition(player : PlayerObject) : void`
Event called when the ball changes position.

**Comment by evi1a** [comment]
Btw is there a way to detect other moving entities despite the default ball and the players atm?

**Comment by Leo-MapCreator** [comment]
The thing is, those function are going to be triggered almostly every tick, why not use onGameTick then?

**Comment by Leo-MapCreator** [comment]
No, you can just detect ball and players, you can't detect map objects moving

**Comment by subxqns** [comment]
how do you detect ball position?

**Comment by grady8** [comment]
u can use room.getBallPosition()

**Comment by Leo-MapCreator** [comment]
You can't set a new position, what you can do is check if the position changed, and let's imagine there are 3 players in both teams, and all of the are moving at the same time, how's that going to happen? It will fire the same function 6 times in the same tick?

**Comment by olegmisar** [comment]
Leo, it's exactly what evi1a asks. And this function will be more complete if it accepts second argument: `onPlayerUpdatePosition(player : PlayerObject, keys: KeysObject) : void`.
`KeysObject` format:
```
up: Triggered by up arrow or "w"
down: Triggered by down arrow or "s"
right: Triggered by right arrow or "d"
left: Triggered by left arrow or "a"
```

**Comment by Leo-MapCreator** [comment]
But i ask you, is that function going to be triggered 6 times in the same thick? Using the example i gave.
Which trigger is from which player, if 6 triggers happens?

**Comment by olegmisar** [comment]
Yes, that function MUST be triggered for EVERY player who changes his position.

**Comment by Leo-MapCreator** [comment]
Honestly, KeysObject dont have **_direct interaction_** with player update position, if you press left when there is a wall on your left you will not move, or even when you are playing bounce that you press down but you are still going up because of the speed you had, doesnt make much sense that

**Comment by Leo-MapCreator** [comment]
What im asking is, how that function is going to be triggered 6 times in the same tick, and how to know which player triggered that? You would still have to detect which player, which is basically doing it without this function

**Comment by olegmisar** [comment]
Computers can do a lot of calculations per tick. Even if there are 100 players in a room and function is called for every player you won't recognize any lag.

**Comment by olegmisar** [comment]
And yeah, you are right about `KeysObject` it makes no sense. So, may be `onPlayerActivity()` should accept event object as argument. I will make an issue.

**Comment by evi1a** [comment]
@olegmisar Actually, it's a great idea to include KeysObject too.
```js
room.onPlayerUpdatePosition = function(player, keys) {
  if(keys == null) {
    // player has been moved
  } else {
    // player moved by himself
  }
}
```

And yes, it would be called everytime a player gets moved. But I don't understand where the problem is. @Leo-MapCreator

**Comment by evi1a** [comment]
@Leo-MapCreator but leo, u are right about **direct interaction** in some cases. if u also include previous position u can count the difference, and say in which direction player actually went.
